spring-cloud-gateway-core/src/main/java/org/springframework/cloud/gateway/filter/factory/RetryGatewayFilterFactory.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.springframework.cloud.gateway.filter.factory;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Predicate;
+
+import reactor.retry.DefaultRepeat;
+import reactor.retry.Repeat;
+import reactor.retry.RepeatContext;
+
+import org.springframework.cloud.gateway.filter.GatewayFilter;
+import org.springframework.cloud.gateway.support.ServerWebExchangeUtils;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatus.Series;
+import org.springframework.tuple.Tuple;
+import org.springframework.util.Assert;
+import org.springframework.web.server.ServerWebExchange;
+
+public class RetryGatewayFilterFactory implements GatewayFilterFactory {
+	@Override
+	public GatewayFilter apply(Tuple args) {
+		Retry retry = new Retry();
+
+		if (args.hasFieldName("retries")) {
+			retry.retries(args.getInt("retries"));
+		}
+
+		// TODO: list of statusSeries
+		if (args.hasFieldName("statusSeries")) {
+			int statusSeries = args.getInt("statusSeries");
+			retry.series(Series.valueOf(statusSeries));
+		}
+
+		// TODO: list of status
+		if (args.hasFieldName("status")) {
+			retry.statuses(ServerWebExchangeUtils.parse(args.getRawString("status")));
+		}
+
+		// TODO: list of methods
+		if (args.hasFieldName("method")) {
+			retry.methods(HttpMethod.resolve(args.getString("method").toUpperCase()));
+		}
+
+		return apply(retry);
+	}
+
+    public GatewayFilter apply(Retry retry) {
+		retry.validate();
+
+		Predicate<? super RepeatContext<ServerWebExchange>> predicate = context -> {
+			ServerWebExchange exchange = context.applicationContext();
+			HttpStatus statusCode = exchange.getResponse().getStatusCode();
+			HttpMethod httpMethod = exchange.getRequest().getMethod();
+
+			boolean retryableStatusCode = retry.getStatuses().contains(statusCode);
+
+			if (!retryableStatusCode) {
+				// try the series
+				retryableStatusCode = retry.getSeries().stream()
+						.anyMatch(series -> statusCode.series().equals(series));
+			}
+
+			boolean retryableMethod = retry.getMethods().contains(httpMethod);
+			return retryableMethod && retryableStatusCode;
+		};
+
+		//TODO: use Repeat statics once updated with a create() like method
+		Repeat<ServerWebExchange> repeat = DefaultRepeat.create(predicate, retry.getRetries());
```

Comment:
you should be able to use `Repeat` builder, by putting the `retry.getRetries()` max attempt condition inside the predicate on line 83. The predicate failing leads to the same signal as the max attempts being reached (see [here](https://github.com/reactor/reactor-addons/blob/master/reactor-extra/src/main/java/reactor/retry/DefaultRepeat.java#L119)).

something like this:
```java
    boolean retryableMethod = retry.getMethods().contains(httpMethod);
    boolean retryableAttempt = context.iteration() < retry.getRetries();
    return retryableMethod && retryableStatusCode && retryableAttempt;
		};
```

Your mileage may vary, the `iteration()` returns a 0-based attempt number, with the original attempt before the first repeat being numbered 0.

Reply:
Thanks, @simonbasle, incorporated.